macros/generic_tests/recency.sql

```diff
@@ -4,7 +4,21 @@
 
 {% macro default__test_recency(model, field, datepart, interval, group_by_columns) %}
 
-{% set threshold = dbt.dateadd(datepart, interval * -1, current_timestamp_backcompat()) %}
+{#- Get columns in model #}
+{%- set columns = adapter.get_columns_in_relation(model) %}
+
+{#- Get column type of field in model #}
+{%- set column_type = columns | selectattr("name", "==", field) | map(attribute="data_type") | first %}
+
+{#- Use a date-based threshold if column type is DATE or datepart is day or greater #}
+{%- if column_type == 'DATE' and datepart in ['year', 'quarter', 'month', 'week', 'day'] %}
+  {%- set threshold = dbt.dateadd(datepart, interval * -1, dbt.safe_cast(dbt.current_timestamp_backcompat(), api.Column.translate_type("date"))) %}
```

Comment:
```suggestion
  {%- set threshold = dbt.dateadd(datepart, interval * -1, dbt.safe_cast(dbt.current_timestamp_backcompat(), 'date')) %}
```

we don't _necessarily_ want `translate_type` to be leaking into user-facing code, but we don't actually have a `dbt.type_date()` macro in Core yet... this one is tricky. 

As far as I can tell, almost all databases call their date-only datatype `date`. The only exception I could easily find was SQLite which either does strings or integers for unix time, which isn't going to work out of the box anyway. 

All this to say I think we're OK to hardcode the string in on this occasion. Lmk if there's something important I've missed!

Reply:
LGTM! For my learning's sake, why is it that we don't want `translate_type` to leak into user-facing code?

Reply:
> why is it that we don't want `translate_type` to leak into user-facing code?

It's better for there to only be one consistent way to do things - if some parts of the package use dbt.type_int() and some use api.Column.translate_type('int') then it makes it a trickier maintenance burden.